File: step_01_basics/1.2_basic_math/IB025_trailing_zeros_in_factorial/solution.py
"""
Problem: Trailing Zeros in Factorial
Difficulty: EASY | XP: 10
Source: InterviewBit

Given an integer n, return the number of trailing zeros in n!.
Trailing zeros are created by pairs of factors 2 and 5.
Since factors of 2 always outnumber factors of 5, count factors of 5 in n!.
"""
from typing import List
import math


# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(n log n)  |  Space: O(n) — factorial grows astronomically
# Compute n! directly then count trailing zeros
# NOTE: Only feasible for very small n; Python big integers handle it but
#       this approach is impractical for n > ~20 in most languages.
# ============================================================
def brute_force(n: int) -> int:
    """
    Actually compute n! using Python's arbitrary-precision integers,
    then count how many times 10 divides it (i.e., count trailing zeros).
    """
    factorial = math.factorial(n)   # Python handles big ints natively
    count = 0
    while factorial % 10 == 0:
        count += 1
        factorial //= 10
    return count


# ============================================================
# APPROACH 2: OPTIMAL
# Time: O(log5 n)  |  Space: O(1)
# Count factors of 5 using Legendre's formula
# ============================================================
def optimal(n: int) -> int:
    """
    Trailing zeros come from factors of 10 = 2 * 5.
    Factors of 2 always exceed factors of 5 in n!, so count factors of 5.
    Multiples of 5 contribute 1 factor, multiples of 25 contribute 2, etc.
    Formula: floor(n/5) + floor(n/25) + floor(n/125) + ...
    """
    count = 0
    power_of_5 = 5
    while power_of_5 <= n:
        count += n // power_of_5
        power_of_5 *= 5
    return count


# ============================================================
# APPROACH 3: BEST
# Time: O(log5 n)  |  Space: O(1)
# Same formula but written more compactly by repeatedly dividing n
# ============================================================
def best(n: int) -> int:
    """
    Equivalent to Optimal but avoids tracking power_of_5 separately.
    Repeatedly divide n by 5, adding the quotient each time.
    This works because:
      floor(n/5) + floor(n/25) + ... = floor(n/5) + floor(floor(n/5)/5) + ...
    """
    count = 0
    while n >= 5:
        n //= 5
        count += n
    return count


if __name__ == "__main__":
    print("=== Trailing Zeros in Factorial ===")
    test_cases = [
        (5, 1),    # 5! = 120
        (10, 2),   # 10! = 3628800
        (25, 6),   # 25 has two 5s
        (100, 24),
        (0, 0),
        (1, 0),
        (30, 7),
    ]
    for n, expected in test_cases:
        b = brute_force(n)
        o = optimal(n)
        be = best(n)
        status = "OK" if b == o == be == expected else "FAIL"
        print(f"n={n:4d} | Brute: {b:3d} | Optimal: {o:3d} | Best: {be:3d} | Expected: {expected:3d} | {status}")

File: step_01_basics/1.2_basic_math/IB025_trailing_zeros_in_factorial/Solution.java
/**
 * Problem: Trailing Zeros in Factorial
 * Difficulty: EASY | XP: 10
 * Source: InterviewBit
 *
 * Given an integer n, return the number of trailing zeros in n!.
 * Trailing zeros = pairs of (2 * 5). Since 2s always exceed 5s, count factors of 5.
 *
 * @author DSA_Nova
 */
import java.math.BigInteger;

public class Solution {

    // ============================================================
    // APPROACH 1: BRUTE FORCE
    // Time: O(n log n)  |  Space: O(n digits) — BigInteger needed for large n
    // Compute n! directly, then count trailing zeros
    // ============================================================
    static class BruteForce {
        /**
         * Compute the full factorial using BigInteger (handles arbitrary precision),
         * then count trailing zeros by repeatedly dividing by 10.
         * Impractical for large n (e.g., n=1000 produces a ~2568-digit number).
         */
        public static int trailingZeroes(int n) {
            BigInteger factorial = BigInteger.ONE;
            for (int i = 2; i <= n; i++) {
                factorial = factorial.multiply(BigInteger.valueOf(i));
            }
            int count = 0;
            BigInteger ten = BigInteger.TEN;
            while (factorial.mod(ten).equals(BigInteger.ZERO)) {
                count++;
                factorial = factorial.divide(ten);
            }
            return count;
        }
    }

    // ============================================================
    // APPROACH 2: OPTIMAL
    // Time: O(log5 n)  |  Space: O(1)
    // Legendre's formula: count factors of 5 at each power of 5
    // ============================================================
    static class Optimal {
        /**
         * Trailing zeros = min(count of 2s, count of 5s) in the prime factorization of n!.
         * 2s are always more plentiful, so count only 5s.
         * n/5 numbers are divisible by 5, n/25 by 25 (contributing an extra 5), etc.
         */
        public static int trailingZeroes(int n) {
            int count = 0;
            long powerOf5 = 5;
            while (powerOf5 <= n) {
                count += n / powerOf5;
                powerOf5 *= 5;
            }
            return count;
        }
    }

    // ============================================================
    // APPROACH 3: BEST
    // Time: O(log5 n)  |  Space: O(1)
    // Compact iterative: repeatedly divide n by 5
    // ============================================================
    static class Best {
        /**
         * Instead of tracking the growing power of 5, divide n by 5 repeatedly.
         * This is equivalent to the Optimal approach mathematically:
         *   floor(n/5) + floor(n/25) + ... == floor(n/5) + floor(n/5 / 5) + ...
         */
        public static int trailingZeroes(int n) {
            int count = 0;
            while (n >= 5) {
                n /= 5;
                count += n;
            }
            return count;
        }
    }

    public static void main(String[] args) {
        System.out.println("=== Trailing Zeros in Factorial ===");
        int[] inputs   = {5, 10, 25, 100, 0, 1, 30};
        int[] expected = {1,  2,  6,  24, 0, 0,  7};

        for (int i = 0; i < inputs.length; i++) {
            int n  = inputs[i];
            int b  = BruteForce.trailingZeroes(n);
            int o  = Optimal.trailingZeroes(n);
            int be = Best.trailingZeroes(n);
            String status = (b == o && o == be && be == expected[i]) ? "OK" : "FAIL";
            System.out.printf("n=%-4d | Brute: %-3d | Optimal: %-3d | Best: %-3d | Expected: %-3d | %s%n",
                    n, b, o, be, expected[i], status);
        }
    }
}
